lib/system/iterators.nim

```diff
@@ -120,13 +120,139 @@ iterator items*[T: enum and Ordinal](E: typedesc[T]): T =
+  ## Iterates over each item of `a` in reverse so that the yielded value can be modified.
+  var i = high(a)
+  while i >= low(a):
+    yield a[i]
+    dec(i)
+
+iterator ritems*[IX, T](a: array[IX, T]): T {.inline.} =
+  ## Iterates over each item of `a` in reverse.
+  when a.len > 0:
+    var i = high(IX)
+    while true:
+      yield a[i]
+      if i < low(IX): break
+      dec(i)
+
+iterator mritems*[IX, T](a: var array[IX, T]): var T {.inline.} =
+  ## Iterates over each item of `a`in reverse so that the yielded value can be modified.
+  when a.len > 0:
+    var i = high(IX)
+    while true:
+      yield a[i]
+      if i < low(IX): break
+      dec(i)
+
+iterator ritems*[T](a: set[T]): T {.inline.} =
+  ## Iterates over each element of `a` in reverse. `items` iterates only over the
+  ## elements that are really in the set (and not over the ones the set is
+  ## able to hold).
+  var i = high(T).int
+  while i > low(T).int:
+    if T(i) in a: yield T(i)
+    dec(i)
+
+iterator ritems*(a: cstring): char {.inline.} =
+  ## Iterates over each item of `a` in reverse.
+  runnableExamples:
+    from std/sequtils import toSeq
+    assert toSeq("abc\0def".cstring) == @['a', 'b', 'c']
+    assert toSeq("abc".cstring) == @['a', 'b', 'c']
+  #[
+  assert toSeq(nil.cstring) == @[] # xxx fails with SIGSEGV
+  this fails with SIGSEGV; unclear whether we want to instead yield nothing
+  or pay a small price to check for `nil`, a benchmark is needed. Note that
+  other procs support `nil`.
+  ]#
+  template impl() =
+    var i = high(a)
+    let n = low(a)
+    while i >= n:
+      yield a[i]
+      dec(i)
+  when defined(js): impl()
+  else:
+    when nimvm:
+      # xxx `cstring` should behave like c backend instead.
+      impl()
+    else:
+      var i = high(a)
+      while a[i] != '\0':
+        yield a[i]
+        dec(i)
+
+iterator mritems*(a: var cstring): var char {.inline.} =
+  ## Iterates over each item of `a` in reverse so that the yielded value can be modified.
+  # xxx this should give CT error in js RT.
+  runnableExamples:
+    from std/sugar import collect
+    var a = "abc\0def"
+    var b = a.cstring
+    let s = collect:
+      for bi in mritems(b):
+        if bi == 'b': bi = 'B'
+        bi
+    assert s == @['a', 'B', 'c']
+    assert b == "aBc"
+    assert a == "aBc\0def"
+
+  template impl() =
+    var i = high(a)
+    let n = low(a)
+    while i >= n:
+      yield a[i]
+      dec(i)
+  when defined(js): impl()
+  else:
+    when nimvm: impl()
+    else:
+      var i = high(a)
+      while a[i] != '\0':
+        yield a[i]
+        dec(i)
+
+iterator ritems*[T: enum and Ordinal](E: typedesc[T]): T =
+  ## Iterates over the values of `E` in reverse.
+  ## See also `enumutils.items` for enums with holes.
+  runnableExamples:
+    type Goo = enum g0 = 2, g1, g2
+    from std/sequtils import toSeq
+    assert Goo.toSeq == [g0, g1, g2]
+  for v in high(E) .. low(E):
+    yield v
```

Comment:
If this gets added, we should probably also have `enumutils.ritems` for enums with holes.

Reply:
> What about only making an iterator called `reverse` which counts down form high to low and yields the element at that index and converting `ritems` and the others to templates that only do `items.reverse`

I just realize this exists https://nim-lang.org/docs/algorithm.html#reverse%2CopenArray%5BT%5D.

Reply:
Done, please check it

Reply:
Looks fine.